plugins/src/lib.rs

```diff
@@ -0,0 +1,190 @@
+use crate::codec::{JsonCodec, JsonRpcCodec};
+use futures::sink::SinkExt;
+use std::sync::Arc;
+use tokio::sync::Mutex;
+use tokio_util::codec::FramedWrite;
+mod codec;
+mod messages;
+pub use anyhow::Error;
+use log::{trace, warn};
+use std::marker::PhantomData;
+use tokio::io::{AsyncRead, AsyncWrite};
+use tokio_stream::StreamExt;
+use tokio_util::codec::FramedRead;
+
+#[macro_use]
+extern crate serde_json;
+
+/// Builder for a new plugin.
+pub struct Builder<S, I, O>
+where
+    S: Clone + Send,
+    I: AsyncRead + Unpin,
+    O: Send + AsyncWrite + Unpin,
+{
+    state: S,
+
+    input: I,
+    output: O,
+    hooks: Hooks,
+    subscriptions: Subscriptions,
+}
+
+impl<S, I, O> Builder<S, I, O>
+where
+    O: Send + AsyncWrite + Unpin + 'static,
+    S: Clone + Send + 'static,
+    I: AsyncRead + Send + Unpin + 'static,
+{
+    pub fn new(state: S, input: I, output: O) -> Self {
+        Self {
+            state,
+            input,
+            output,
+            hooks: Hooks::default(),
+            subscriptions: Subscriptions::default(),
+        }
+    }
+
+    #[allow(unused_mut)]
+    pub async fn run(mut self) -> Plugin<S, I, O> {
+        let (plugin, input) = self.build();
+        tokio::task::spawn(plugin.run(input));
+        tokio::time::sleep(std::time::Duration::from_secs(300)).await;
+        unimplemented!()
```

Comment:
Refused?

Reply:
Yeah, it's work in progress, and therefore in draft mode (not meant to be reviewed yet). That will be fixed before undrafting